### PR TITLE
Dionisia - Edges - Hotel

### PR DESCRIPTION
**Pull request by larachan15**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One decision I made was to not use a module. I started to create one, but then it didn’t make sense to me because this is a small project and I only had 3 classes. I also wasn’t sure if I should make a room class or not. But in the end, because I didn’t have a module, I decided to create the room class to try and make my 3 classes have as much single responsibility as possible.
Describe a concept that you gained more clarity on as you worked on this assignment. | I am still struggling with proper syntax. I had a LOT of errors in my code and so I’m becoming a lot better at reading through the stack trace. I also gained more clarity writing the tests. I definitely need to write more tests and this became a lot easier to do towards the end of the weekend.
Describe a _nominal test_ that you wrote for this assignment. | I wrote a test to check to see if reservation_tracker calculated the cost_of_a_given_reservation properly
Describe an _edge case test_ that you wrote for this assignment. | I wrote a test to check if a check_out_date was entered in before a check_in_date for reservation
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think I need more practice with this. I didn’t write much psuedocode. I tried to diagram the relationships between the methods and the classes that I thought I needed and then wrote tests. I did switch into writing code first and then creating some tests after because I was running out of time and that process for me is still faster at this time.

### CODE REVIEW

**Comment by tildeee** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| x
Answer comprehension questions	| x
**Design** |
Each class is responsible for a single piece of the program | x
Classes are loosely coupled | x
**Wave 1** |
List rooms | x
Reserve a room for a given date range | x
List reservations for a given date | x
Calculate reservation price | x
Invalid date range produces an error | x
Test coverage | x
**Wave 2** |
View available rooms for a given date range | x
Reserving a room that is not available produces an error | ? maybe I'm not seeing if this requirement is met
Test coverage | x... I expected more tests around date conflicts when creating reservations and for a lot more of your methods.
**Wave 3** |
Create a block of rooms | x
Check if a block has rooms | x
Reserve a room from a block | 
Test coverage | getting there...
**Fundamentals** | 
Names variables, classes and modules appropriately | x
Understanding of variable scope - local vs instance | x
Can create complex logical structures utilizing variables | x yes!
Appropriately uses methods to break down tasks into smaller simpler tasks | x
Understands the differences between class and instance methods | x
Appropriately uses iterators and enumerables | x, you took ownership of this!
Appropriately writes and utilizes classes | x
Appropriately utilizes modules as a mechanism to encapsulate functionality | Didn't use a module in this project, and that's okay!
**Wrap Up** |
There is a refactors.txt file | 
The file provides a roadmap to future changes | 
**Additional Feedback** | 

Dionisia! You did a great job with this project. I think the code is really clean and logical -- I am impressed by how slim ReservationTracker turned out, and I'm pleased with that! I think that your ability to feel comfortable in Ruby (using `map`, iteration, `pop`/array manipulation, classes, constants) contributed to how nice the code feels.

I'm adding a bunch of comments that I think will amplify... _elevate_ ... your code style.

Overall, I feel that your tests are lacking. I think you acknowledge that based on your comprehension question answers, but I just wanted to say it out loud! It's just a shame because I think the code style is great, but I can't be confident that it's working as required because of the lack of tests.

Otherwise, the code looks great... If the code actually works accurately, then the code looks really great. To be quite clear: I don't trust that much of this code works as you'd like it to/there are more bugs than you think! But code-wise and design-wise, I think this is a great start. Good job!

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+require 'securerandom'
+
+class Block
+  attr_reader :check_in_date, :check_out_date, :id, :block_of_rooms
+  attr_accessor :rooms
+
+  BLOCKRATE = 100
+
+  def initialize(check_in_date, check_out_date, block_of_rooms)
+    @id = SecureRandom.uuid
```
Comment (tildeee) [inline lib/block.rb:11]: I think your use of the `SecureRandom` gem here to generate an ID is rad!

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+ROOM_LIST = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
+
+def hotel
+  return "hotel"
+end
```
Comment (tildeee) [inline lib/hotel.rb:5]: Whoops! You should probably delete this file and all references to it.

File: lib/reservation.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# module Hotel
+class Reservation
+  attr_reader :check_in_date, :check_out_date, :room_number, :nights_stayed
+
+  def initialize(check_in_date, check_out_date, room_number)
+    @check_in_date = check_in_date
+    @check_out_date = check_out_date
+    @room_number = room_number
+    raise StandardError, 'The end date cannot be before the start date.' if check_out_date <= check_in_date
```
Comment (tildeee) [inline lib/reservation.rb:13]: I love that the logic to check if reservation dates are valid are in the Reservation itself. Nice decision :)

File: lib/reservation.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Keeps track of 1 reservation
+require_relative 'room'
+require 'date'
+
+# module Hotel
+class Reservation
+  attr_reader :check_in_date, :check_out_date, :room_number, :nights_stayed
```
Comment (tildeee) [inline lib/reservation.rb:7]: `attr_reader`s are reserved for names of instance variables. `:nights_stayed` isn't the name of an instance var, it's the name of a method -- it doesn't need to be in this list!

File: lib/reservation_tracker.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+require_relative 'reservation'
+# Keeps track of the list of reservations
+# module Hotel
+class ReservationTracker
+  attr_reader :rooms, :check_in_date, :check_out_date, :reservations, :unreserved_rooms, :blocks
+
+  def initialize
+    @rooms = create_rooms
+    @reservations = []
+    @unreserved_rooms = []
```
Comment (tildeee) [inline lib/reservation_tracker.rb:14]: Does this instance variable `@unreserved_rooms` ever get used? If it doesn't get used, can we get rid of it?

File: lib/reservation_tracker.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  #   @reservations.each do |reservation |
+  #     return false if reservation.check_in_date >= date && reservation.check_out_date > date
+  #   end
+  #   # binding.pry
+  #   return true
+  # end
+
+  # accesses the list of reservations for a specific date
+  def list_of_reservations(date)
+    return @reservations.find_all { |reservation| reservation.nights_stayed == date }
```
Comment (tildeee) [inline lib/reservation_tracker.rb:63]: Does this code actually work? It looks fishy to me, and I can't tell because there aren't any tests! :(

File: lib/reservation_tracker.rb [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    total_cost = reservation.nights_stayed * Room::ROOM_RATE
+    return total_cost
+  end
+
+  # checks to see if a room is available and reserves the first available room for a given date range. Uses the rooms_not reserved method.
+  def reserve_room(check_in_date, check_out_date)
+     # set the rooms_not_reserved method and arguments to a local variable.
+     unreserved_rooms = rooms_not_reserved(check_in_date, check_out_date)
+     if unreserved_rooms.length == 0
+      raise ArgumentError, 'No Available Rooms for Given Dates'
```
Comment (tildeee) [inline lib/reservation_tracker.rb:77]: Instead of raising an ArgumentError, it might make more sense to make your own custom error or just raise `StandardError`!

File: lib/reservation_tracker.rb [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+      create_reservation(check_in_date, check_out_date, unreserved_rooms[0])
+      # binding.pry
+    end
+  end
+
+  # accesses a list of rooms that are not reserved for a given date range
+  def rooms_not_reserved(check_in_date, check_out_date)
+    # if the reservations array is empty, this returns an array of available room numbers. If there are no reservations, all rooms can be reserved.
+    return @rooms.map { |room| room.room_number } if @reservations.empty?
+    unreserved_rooms = @rooms.map { |room| room.room_number }
```
Comment (tildeee) [inline lib/reservation_tracker.rb:89]: Love how dense and juicy and exciting the above two lines are!

File: lib/reservation_tracker.rb [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  # calculates the total cost for a given reservation
+  def cost_for_given_reservation(reservation)
+    total_cost = reservation.nights_stayed * Room::ROOM_RATE
+    return total_cost
+  end
+
+  # checks to see if a room is available and reserves the first available room for a given date range. Uses the rooms_not reserved method.
+  def reserve_room(check_in_date, check_out_date)
+     # set the rooms_not_reserved method and arguments to a local variable.
+     unreserved_rooms = rooms_not_reserved(check_in_date, check_out_date)
```
Comment (tildeee) [inline lib/reservation_tracker.rb:75]: I love this line! Very clear what's going on here.

File: lib/reservation_tracker.rb [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+  # accesses a list of rooms that are not reserved for a given date range
+  def rooms_not_reserved(check_in_date, check_out_date)
+    # if the reservations array is empty, this returns an array of available room numbers. If there are no reservations, all rooms can be reserved.
+    return @rooms.map { |room| room.room_number } if @reservations.empty?
+    unreserved_rooms = @rooms.map { |room| room.room_number }
+    # if @reservation.empty? is not empty
+    @reservations.each { |reservation|
+      if (reservation.check_in_date >= check_in_date && reservation.check_in_date <= check_out_date) || (reservation.check_out_date >= check_in_date && reservation.check_out_date <= check_out_date)
+        unreserved_rooms.delete(reservation.room_number)
```
Comment (tildeee) [inline lib/reservation_tracker.rb:93]: This is a very funny assumption you're putting here:

For the call `.delete` to work, this assumes that the array `unreserved_rooms` is an array full room numbers (originally stemming from `@rooms`) that are _unique_. If they aren't unique, then this call to `delete` might not do what you expect... so if the requirement changes that the hotel does not have unique room numbers, this may introduce bugs here.

File: lib/reservation_tracker.rb [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      end
+    }
+
+    @blocks.each { |block|
+        if (block.check_in_date >= check_in_date &&  block.check_in_date <= check_out_date) || (block.check_out_date >= check_in_date && block.check_out_date <= check_out_date)
+          unreserved_rooms - block.rooms
```
Comment (tildeee) [inline lib/reservation_tracker.rb:99]: This is cool code here! I like it. However... the `-` isn't destructive. It just does the subtraction, and the result isn't stored anywhere. Observe in irb:
```ruby
a = [1,2,3]
b = [1,2]
a - b
# => returns [3], as expected
# a is still [1,2,3]
# b is still [1,2]
```

So the subtraction doesn't do anything!

File: lib/reservation_tracker.rb [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+  def cost_for_given_reservation(reservation)
+    total_cost = reservation.nights_stayed * Room::ROOM_RATE
+    return total_cost
+  end
+
+  # checks to see if a room is available and reserves the first available room for a given date range. Uses the rooms_not reserved method.
+  def reserve_room(check_in_date, check_out_date)
+     # set the rooms_not_reserved method and arguments to a local variable.
+     unreserved_rooms = rooms_not_reserved(check_in_date, check_out_date)
+     if unreserved_rooms.length == 0
```
Comment (tildeee) [inline lib/reservation_tracker.rb:76]: The indentation here is off ;)
